### PR TITLE
fix(votes): stop empty table on uncached page jumps (LFXV2-3199)

### PR DESCRIPTION
**Pull request by audigregorie**
## Summary

The votes dashboard uses cursor-based pagination, where the API only ever returns the token for the *next* page. Previously, jumping directly to a page whose token had not yet been collected rendered an empty table and silently reset the pager to page 1. This PR replaces that guard with a forward walk from the nearest cached token, so direct page jumps work at any depth.

Resolves LFXV2-3199

## Behavior changes

| Before | After |
|---|---|
| Clicking a page number in the votes table that had never been visited (e.g. jumping from page 1 straight to page 5) showed an empty table and snapped the pager back to page 1, with no way to reach that page except clicking "next" repeatedly. | Jumping to any page walks forward through the cursor chain behind the scenes and displays the requested page. If the data has shrunk since the page count was fetched (e.g. votes were deleted), the pager gracefully falls back to page 1 instead of showing an empty table. |

## Technical changes

`apps/lfx-one/src/app/modules/votes/votes-dashboard/votes-dashboard.component.ts` — Votes dashboard pagination
- Replaces the page-token guard (which reset `currentFirst` to 0 and returned an empty list for uncached pages) with a new private `fetchVotePage()` method.
- `fetchVotePage()` finds the nearest page fetchable directly (page 0 needs no token; page p needs `pageTokens[p - 1]`) and recursively walks forward via `switchMap`, caching each intermediate page token so later jumps are cheaper.
- Handles stale `totalRecords` (cursor exhausts before the target page): falls back to page 1, reusing the already-fetched page-0 response when the walk started there to avoid a duplicate request.
- The `initVotes()` pipeline now delegates to `fetchVotePage()` and maps the response to `response.data`; loading/error handling is unchanged.
- Updates the constructor comment on the lens-change pagination reset to reference the walk instead of the removed guard.

### CODE REVIEW

File: apps/lfx-one/src/app/modules/votes/votes-dashboard/votes-dashboard.component.ts [hunk 1]
```diff
@@ -271,40 +271,65 @@ export class VotesDashboardComponent {
...
+      fetchSingle(index).pipe(
+        switchMap((response: PaginatedResponse<Vote>) => {
+          if (index === pageIndex) {
+            return of(response);
+          }
+          if (!response.page_token) {
+            this.currentFirst.set(0);
+            return index === 0 ? of(response) : fetchSingle(0);
+          }
+          return walk(index + 1);
```
[the remark this answers is not in the document]
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/votes-dashboard/votes-dashboard.component.ts]: Fixed in 9fcbed6a6 — the target-page return now treats an empty nonzero page with no `page_token` as cursor exhaustion: the stale token chain is cleared, the paginator resets to page 1, and page 1 is refetched. The mid-walk exhaustion branch now clears the token chain too, so stale tokens beyond an exhausted page are never reused.

**Review by luismoriguerra — DISMISSED** [review]
## Audit: PR #1417 — fix(votes): stop empty table on uncached page jumps (LFXV2-3199)

`feat/lfxv2-3199-votes-page-jump-fix` → `main` · 1 file, +63/−27 · head `c76070ff`

**Scope:** single Angular component — cursor-pagination page-jump logic in `votes-dashboard.component.ts`.

**Lanes run:** gates re-run in an isolated worktree (install, `lint:check`, `format:check`, `check-types`, `build`) · secrets/critical-constants · code-standards.
**Skipped:** architecture/ADR and platform-contract-coherence (no v2 write→read surface — no Goa design, RuleSet/HTTPRoute, NATS subject, KV or FGA change); privacy/PII (no new field on any payload, no new read/write path — same `/api/votes` call, same params).

### 1. Why this change

Votes uses cursor pagination and the API returns only the *next* page's token. The old guard returned `of([])` and silently reset `currentFirst` to 0 whenever `pageTokens[pageIndex - 1]` was missing — so clicking page 5 from page 1 showed an empty table and snapped back. Verified against the removed code.

### 2. How it was done

New private `fetchVotePage()`: finds the nearest directly-fetchable page (`startIndex`), then `walk()` recursively fetches forward via `switchMap`, caching intermediate tokens. Exhaustion mid-walk or an empty target page clears the token chain and falls back to page 1. `initVotes()` delegates and maps to `response.data`; loading/error handling unchanged.

Both prior bot findings (Copilot on L328, cursor[bot] "stale tokens poison page walks" on L339) are **genuinely fixed** in `9fcbed6a` / `c76070ff` — confirmed in code at the `pageTokens.length = index` truncation and the empty-target-page exhaustion branch.

### 3. Is it correct

Yes for the stated bug. Gates re-run clean:

- `yarn lint:check` → 0 errors, 1 warning (`campaigns.component.ts:59`, pre-existing, unrelated file)
- `yarn format:check` → clean
- `yarn check-types` → clean
- `yarn build` → succeeds (48s)
- No unit tests exist for this or any sibling paginated dashboard, so nothing to run.

Reset paths verified: `pageTokens` is cleared on lens change, refresh, rows-per-page change, and all three filter/search handlers — so tokens never leak across a different query. `[first]="currentFirst()"` is bound in the template, so the in-stream reset is reflected in the paginator.

### 4. Findings

No blocking findings. The following are reported here rather than opened as inline threads:

1. **Non-blocking — unbounded serial fetch on a cold cache.** `showFirstLastIcon` defaults `true` (`table.component.ts:74`), so one click on "last page" with an empty `pageTokens` (after any refresh / filter / lens change) issues `N` sequential round-trips with no cap and no incremental feedback. The repo's only other cursor walk bounds itself — `meetings-dashboard.component.ts:912-914` uses `expand(...)` + `take(10)`. Suggest a cap with a defined overflow behavior (clamp to the last reachable page rather than silently resetting to page 1).

2. **Non-blocking — a transient HTTP error is indistinguishable from cursor exhaustion.** `VoteService.getVotes` (`vote.service.ts:29-33`) swallows all errors into `{ data: [], page_token: undefined }`. A 500 mid-walk therefore hits `pageTokens.length = index` (truncating the cache) and then `pageTokens = []` + `currentFirst.set(0)` + refetch page 0 — the user is silently bounced to page 1 by a network blip. Self-recovering, and no worse *visually* than the old empty table, but the diff newly destroys pagination state on error. Fix direction: surface a discriminated failure from the service (or bypass the swallowing wrapper here) so exhaustion and failure branch differently.

3. **Non-blocking — divergence from the repo idiom.** `walk()` hand-rolls `expand()`, nesting one `switchMap` layer per page walked. `meetings-dashboard.component.ts:912` uses `expand()`, which flattens the chain and makes the bound explicit.

4. **Nit** — the doc comment on `fetchVotePage` and the inline comment on the target-page branch assert "the cached cursor was stale" for a condition the code cannot distinguish from a failed request (see #2).

5. **Nit** — no unit test for a ~45-line branching method with mutable side effects (`docs/reviews/knowledge-base/code-truthiness.md:135`, `non-trivial-logic-without-unit`). The exhaustion/fallback branches are exactly what nobody exercises manually. No sibling dashboard has a spec either, so this is precedent-consistent rather than a hard convention break.

6. **Nit** — three mutation idioms on `pageTokens` (index-assign, `length =` truncate, full reassign) across six reset sites. A small token-cache helper would make the invariant "index `i` holds the token for page `i+1`" enforceable and unit-testable.

7. **🔒 Nit** — no secrets and no critical-constant drift: `rowsPerPage` (10), endpoints, params and error handling are all unchanged.

### 5. Q&A this PR resolves

- **Q:** Can a stale token now render a wrong page's data silently? **A:** Only if the stale cursor still returns rows. Empty-with-no-token is caught; a stale-but-nonempty cursor is not detectable client-side and is accepted as a trade-off.
- **Q:** Does `currentFirst.set(0)` inside the `switchMap` re-trigger the pipeline? **A:** No — `currentFirst` is read imperatively, not via `toObservable` in the `combineLatest`, so there is no feedback loop today. The coupling is implicit and one refactor away from becoming one.
- **Q:** Is `totalRecords` refreshed on the page-1 fallback? **A:** No; the count comes from a separate pipeline, so the paginator can briefly show a stale page count after a shrink-triggered fallback. Cosmetic.

### 6. Business rules & ADRs

- **Business rule:** any page reachable per `totalRecords` must be directly reachable by the paginator; when the underlying data has shrunk, degrade to page 1 rather than an empty table.
- **ADR alignment:** n/a — pure client-side UI change with no identity, transport, data-store, auth-gating or migration surface.
- **Privacy/access note:** checked — the diff adds no field to any response payload and introduces no new read/write path; it reuses `getVotesByProjectPaginated` with identical params and the same server-side authorization. No PII surface change.

**Verdict:** ✅ Approved
(merge readiness: **merge-after-confirming** the unbounded "last page" walk in finding #1 is acceptable for the largest expected vote count per project — that is the one finding a user can feel, and the sibling implementation deliberately caps it.)

---

_Audited with `pr-reviewer-agent`. No blocking findings, so the summary is posted without inline threads; the 3 non-blocking findings and 4 nits above were not opened as separate comment threads._

**Review by MRashad26 — DISMISSED** [review]
Reviewed with the code-standards-enforcer checklist (Angular/RxJS conventions, template rules, shared interfaces/constants placement).

**Summary**: Replaces the page-token guard (which reset to page 1 and rendered an empty table on uncached direct page jumps) with a `fetchVotePage()` recursive forward-walk from the nearest cached cursor token, per LFXV2-3199.

**Correctness trace**:
- `startIndex` search correctly finds the nearest directly-fetchable page ≤ target (checks `pageTokens[pageIndex-1]` first, so an already-cached target page short-circuits with no walk at all).
- The walk correctly distinguishes a genuinely-last page with data from a stale/exhausted cursor: the "reset to page 1" branch only fires on `!page_token && !data.length`, and only that check is target-page-specific — an *intermediate* page with no next token is unambiguously exhaustion regardless of its own data, which the code handles right.
- Avoids a duplicate request when the walk already started at page 0 during a fallback (`index === 0 ? of(response) : fetchSingle(0)`), matching the PR description.
- `pageTokens.length = index` correctly prunes stale cached tokens beyond the real last page after data shrinks.

**Other notes**:
- No functions introduced in templates (HTML untouched), no new interfaces/constants to misplace — nothing to flag there.
- No unit test file exists for this component at all (pre-existing gap, not introduced by this PR), so this recursive/edge-case-heavy logic is currently only exercised by manual/e2e testing if any — worth a follow-up ticket for spec coverage given the branching complexity, but not a blocker for this fix.

No violations found. Approving.

**Review by ahmedomosanya — DISMISSED** [review]
LGTM — the cursor walk fixes uncached page jumps, and the follow-ups correctly clear stale tokens and fall back to page 1 when the chain exhausts. No blockers from my pass.

**Review by MRashad26 — APPROVED** [review]

**Review by luismoriguerra — APPROVED** [review]
## Audit: PR #1417 — fix(votes): stop empty table on uncached page jumps (LFXV2-3199)
audigregorie · `feat/lfxv2-3199-votes-page-jump-fix` → `main` · MERGED · 5 files, +162/−33 · head `289558d6`

**Verdict:** ✅ Approved
(merge readiness: already-merged)

Gates (isolated worktree): `yarn format:check` clean · `yarn lint:check` 0 errors (1 pre-existing warning in `campaigns.component.ts:59`) · `yarn check-types` clean · `yarn build` succeeded · `@lfx-one/shared` vitest 730/730.

No blocking findings, so this summary is posted without inline threads. 4 non-blocking findings and 4 nits are covered here but were not opened as separate comment threads. Say `all` to also anchor them inline.

### Findings
1. **Non-blocking** — `packages/shared/src/utils/vote.utils.ts:233` · no spec for `recordPageToken` / `findCursorWalkStartIndex` / `resolveCursorWalkOutcome` · add `vote.utils.spec.ts`.
2. **Non-blocking** — `docs/architecture/backend/pagination.md:99` · still documents the old token-guard and `getVotes` wrapper · update to walk/clamp/restart.
3. **Non-blocking** — `apps/lfx-one/src/app/modules/votes/votes-dashboard/votes-dashboard.component.ts:328` · exhaustion restart does not refresh `totalRecords` · call `refresh$.next()` on restart.
4. **Non-blocking** — `packages/shared/src/utils/vote.utils.ts:233` · generic cursor helpers live in vote/poll files · move to `pagination.utils.ts` / `api.interface.ts`.
5. **Nit** — `votes-dashboard.component.ts:319` · `take(VOTES_PAGE_WALK_LIMIT + 1)` is 11 requests vs meetings `take(10)` · fix the “matches meetings” comment.
6. **Nit** — `packages/shared/src/constants/poll.constants.ts:24` · JSDoc cites “last page”; votes table sets `[showFirstLastIcon]="false"`.
7. **Nit** — `votes-dashboard.component.ts:17` · `@lfx-one/shared/utils` import sits after `@services` · group with other `@lfx-one/shared*` imports.
8. **🔒 Nit** — no secrets; `VOTES_PAGE_WALK_LIMIT = 10` is intentional.

_Audited with `pr-reviewer-agent`._
